sound/soc/sof/ipc3-topology.c

```diff
@@ -2003,10 +2003,13 @@ static int sof_ipc3_dai_config(struct snd_sof_dev *sdev, struct snd_sof_widget *
 
 	config->flags = flags;
 
-	ret = sof_ipc_tx_message(sdev->ipc, config, config->hdr.size,
-				 &reply, sizeof(reply));
-	if (ret < 0)
-		dev_err(sdev->dev, "Failed to set dai config for %s\n", dai->name);
+	/* only send the IPC if the widget is set up in the DSP */
+	if (swidget->use_count > 0) {
+		ret = sof_ipc_tx_message(sdev->ipc, config, config->hdr.size,
+					 &reply, sizeof(reply));
+		if (ret < 0)
+			dev_err(sdev->dev, "Failed to set dai config for %s\n", dai->name);
+	}
```

Comment:
@ranj063, this patch broke audio on my laptop:
```
aplay -i -Dplughw:0,0 /home/sof/work/samples/Slayer-Silent_Scream-192K_24bit.wav
```
```
[  411.219537] sof-audio-pci-intel-tgl 0000:00:1f.3: fw_state change: 5 -> 6
[  411.219555] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x80]=0x40000 successful
[  411.219579] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x80]=0x40000 successful
[  411.219585] sof-audio-pci-intel-tgl 0000:00:1f.3: hda_dsp_stream_setup_bdl: period_bytes:0x0
[  411.219587] sof-audio-pci-intel-tgl 0000:00:1f.3: hda_dsp_stream_setup_bdl: periods:1
[  411.219602] sof-audio-pci-intel-tgl 0000:00:1f.3: snd_sof_enable_trace: stream_tag: 1
[  411.219615] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx: 0x90030000: GLB_TRACE_MSG: DMA_PARAMS_EXT
[  411.220016] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG: DMA_POSITION
[  411.220053] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx succeeded: 0x90030000: GLB_TRACE_MSG: DMA_PARAMS_EXT
[  411.220064] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG: DMA_POSITION
[  411.220088] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x80]=0x14001e successful
[  411.220116] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx: 0x40020000: GLB_PM_MSG: CTX_RESTORE
[  411.220323] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx succeeded: 0x40020000: GLB_PM_MSG: CTX_RESTORE
[  411.261524] sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: open stream 0 dir 0
[  411.261532] sof-audio-pci-intel-tgl 0000:00:1f.3: period min 192 max 16384 bytes
[  411.261535] sof-audio-pci-intel-tgl 0000:00:1f.3: period count 2 max 16
[  411.261537] sof-audio-pci-intel-tgl 0000:00:1f.3: buffer max 65536 bytes
[  411.262220] sof-audio-pci-intel-tgl 0000:00:1f.3: hda_ctrl_dai_widget_setup: DAI config failed for widget HDA0.OUT
[  411.262227] sof-audio-pci-intel-tgl 0000:00:1f.3: ASoC: error at snd_soc_dai_hw_params on Analog CPU DAI: -2047376832
[  411.262303]  Analog Playback and Capture: ASoC: __soc_pcm_hw_params() failed (-2047376832)
[  411.262307]  HDA Analog: ASoC: dpcm_fe_dai_hw_params failed (-2047376832)
[  411.262315] sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: free stream 0 dir 0
[  411.262596] sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: close stream 0 dir 0
[  411.719980] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG: DMA_POSITION
[  411.720030] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG: DMA_POSITION
[  412.220094] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG: DMA_POSITION
[  412.220140] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG: DMA_POSITION
[  413.762652] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x80]=0x20140000 successful
[  413.762686] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx: 0x90050000: GLB_TRACE_MSG: DMA_FREE
[  413.762933] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx succeeded: 0x90050000: GLB_TRACE_MSG: DMA_FREE
[  413.790923] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx: 0x40010000: GLB_PM_MSG: CTX_SAVE
[  413.791161] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx succeeded: 0x40010000: GLB_PM_MSG: CTX_SAVE
[  413.791188] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x4]=0x3010303 successful
[  413.791713] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x4]=0x303 successful
[  413.791718] sof-audio-pci-intel-tgl 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask 1
[  413.791787] sof-audio-pci-intel-tgl 0000:00:1f.3: Debug PCIR: 00000010 at  00000044
[  413.791794] sof-audio-pci-intel-tgl 0000:00:1f.3: Current DSP power state: D3
[  413.791795] sof-audio-pci-intel-tgl 0000:00:1f.3: fw_state change: 6 -> 0
```
Bisect log on topic/sof-dev:
```
git bisect start
# bad: [f12c84ec5e558d250f8a35c7527286096014e74d] fixup! ASoC: SOF: topology: Make route setup IPC agnostic
git bisect bad f12c84ec5e558d250f8a35c7527286096014e74d
# good: [f1e2b3be6f5f16a3b932eb56d453084d9ae4bd4c] Fixup! ASoC: SOF: Add a new op to set up volume table
git bisect good f1e2b3be6f5f16a3b932eb56d453084d9ae4bd4c
# bad: [06d92522800a2bae6cc7e393fb858b9f8135ad1e] ASoC: SOF: ipc3: Add local implementation for handling fw_ready message
git bisect bad 06d92522800a2bae6cc7e393fb858b9f8135ad1e
# good: [74be7245d2590f950a9be3f12dc6e011f276404c] fixup! ASoC: SOF: topology: Add ops for setting up and tearing down pipelines
git bisect good 74be7245d2590f950a9be3f12dc6e011f276404c
# bad: [6d6ca220be296cc4ec0313eb1bfd299cb51585f1] fixup! ASoC: SOF: Make sof_widget_setup/free IPC agnostic
git bisect bad 6d6ca220be296cc4ec0313eb1bfd299cb51585f1
# good: [b2473cebfc4d388f4f2dc40fd914a3e981130441] fixup! ASoC: SOF: Add a new dai_get_clk topology IPC op
git bisect good b2473cebfc4d388f4f2dc40fd914a3e981130441
# first bad commit: [6d6ca220be296cc4ec0313eb1bfd299cb51585f1] fixup! ASoC: SOF: Make sof_widget_setup/free IPC agnostic
```

Reply:
Interestingly on up-xtreme i11 it is OK, but not on my tgl-h HDA laptop.

Reply:
OK, up-xtreme i11 no longer works with updated firmwre :o

Reply:
fixed by https://github.com/thesofproject/sof/pull/5347